tools/cldr-code/src/main/java/org/unicode/cldr/util/NameGetter.java

```diff
@@ -0,0 +1,547 @@
+package org.unicode.cldr.util;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.ibm.icu.text.MessageFormat;
+import com.ibm.icu.text.Transform;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class NameGetter {
+
+    private final CLDRFile cldrFile;
+
+    /**
+     * Construct a new NameGetter.
+     *
+     * @param cldrFile must not be null
+     */
+    public NameGetter(CLDRFile cldrFile) {
+        if (cldrFile == null) {
+            throw new RuntimeException("NameGetter must have non-null CLDRFile");
+        }
+        this.cldrFile = cldrFile;
+    }
+
+    private static final String GETNAME_LOCALE_SEPARATOR =
+            "//ldml/localeDisplayNames/localeDisplayPattern/localeSeparator";
+    private static final String GETNAME_LOCALE_PATTERN =
+            "//ldml/localeDisplayNames/localeDisplayPattern/localePattern";
+    private static final String GETNAME_LOCALE_KEY_TYPE_PATTERN =
+            "//ldml/localeDisplayNames/localeDisplayPattern/localeKeyTypePattern";
+
+    private static final Joiner JOIN_HYPHEN = Joiner.on('-');
+    private static final Joiner JOIN_UNDERBAR = Joiner.on('_');
+
+    /**
+     * Get a name, given a type (as a String) and a code.
+     *
+     * <p>Ideally getNameFromTypestrCode and getNameFromTypenumCode should be replaced by a method
+     * that takes an enum rather than String or int. Compare StandardCodes.CodeType, and the
+     * integers defined in the vicinity of CLDRFile.TERRITORY_NAME
+     *
+     * @param type a string such as "currency", "language", "region", "script", "territory", ...
+     * @param code a code such as "JP"
+     * @return the name
+     */
+    public String getNameFromTypestrCode(String type, String code) {
+        return getNameFromTypenumCode(CLDRFile.typeNameToCode(type), code);
```

Comment:
Once you get this merged, a second refactoring is to inline this and similar ones. 'Typenum' is hard to read; TypeEnum is better.

Reply:
Yes, I'll address that in a subsequent PR. By "inline", do you mean, for example, to change `getNameFromTypestrCode("territory", territory)` to `getNameFromTypenumCode(CLDRFile.TERRITORY_NAME, territory)`? And then subsequently (or in the same PR) change CLDRFile.TERRITORY_NAME, etc., to an enum?